src/lib.rs

```diff
@@ -277,532 +135,100 @@ impl TsunamiBuilder {
-                        _ => {
-                            all_ready = false;
-                        }
-                    }
-                }
-            }
-        }
-
-        let mut res = None;
-        let mut errors = Vec::new();
-        let running: u32 = machines.values().map(|ms| ms.len() as u32).sum();
-        if running == expected_num {
-            info!(log, "all machines instantiated; running setup");
-
-            //    - once an instance is ready, run setup closure
-            let usernames = &usernames;
-            let private_key_file = &private_key_file;
-            let wait_for = self.max_wait.map(|wl| wl - start.elapsed());
-            errors.par_extend(machines.par_iter_mut().flat_map(|(name, machines)| {
-                let f = &setup_fns[name];
-                machines
-                    .par_iter_mut()
-                    .map(move |machine| -> Result<_, Error> {
-                        use std::net::{IpAddr, SocketAddr};
-                        let mut sess = ssh::Session::connect(
-                            &log,
-                            &usernames[name],
-                            SocketAddr::new(
-                                machine
-                                    .public_ip
-                                    .parse::<IpAddr>()
-                                    .context("machine ip is not an ip address")?,
-                                22,
-                            ),
-                            private_key_file.path(),
-                            wait_for,
-                        ).context(format!(
-                            "failed to ssh to {} machine {}",
-                            name, machine.public_dns
-                        ))
-                            .map_err(|e| {
-                                error!(log, "failed to ssh to {}:{}", &name, &machine.public_ip);
-                                e
-                            })?;
+            })
+            .try_reduce(Vec::new, |mut provs, prov| {
+                provs.extend(prov);
+                Ok(provs)
+            })?;
 
-                        debug!(log, "setting up {} instance", name; "ip" => &machine.public_ip);
-                        f(&mut sess)
-                            .context(format!("setup procedure for {} machine failed", name))
-                            .map_err(|e| {
-                                error!(log, "setup for {} machine failed", name);
-                                e
-                            })?;
-                        info!(log, "finished setting up {} instance", name; "ip" => &machine.public_ip);
+        Ok(Tsunami(providers, log))
+    }
+}
 
-                        machine.ssh = Some(sess);
-                        Ok(())
+/// When this is dropped, the instances are all terminated automatically.
+///
+/// # Note
+/// See caveats for Azure and Baremetal machines.
+#[must_use]
+pub struct Tsunami<L: Launcher>(Vec<L>, slog::Logger);
```

Comment:
Probably best to name this fields. Took me a while to realize that the `Vec<L>` is a list of one `Launcher` instance per region.